extensions/ql-vscode/src/local-queries.ts

```diff
@@ -7,384 +7,575 @@ import {
-  progress: ProgressCallback,
-  token: CancellationToken,
-  databaseItem: DatabaseItem | undefined,
-  range?: Range,
-): Promise<void> {
-  if (qs !== undefined) {
     // If no databaseItem is specified, use the database currently selected in the Databases UI
     databaseItem =
-      databaseItem || (await databaseUI.getDatabaseItem(progress, token));
+      databaseItem || (await this.databaseUI.getDatabaseItem(progress, token));
     if (databaseItem === undefined) {
       throw new Error("Can't run query without a selected database");
     }
-    const databaseInfo = {
-      name: databaseItem.name,
-      databaseUri: databaseItem.databaseUri.toString(),
-    };
+
+    const coreQueryRun = this.queryRunner.createQueryRun(
+      databaseItem.databaseUri.fsPath,
+      {
+        queryPath: selectedQuery.queryPath,
+        quickEvalPosition: selectedQuery.quickEvalPosition,
+      },
+      true,
+      getOnDiskWorkspaceFolders(),
+      this.queryStorageDir,
+      undefined,
+      undefined,
+    );
 
     // handle cancellation from the history view.
     const source = new CancellationTokenSource();
-    token.onCancellationRequested(() => source.cancel());
-
-    const initialInfo = await createInitialQueryInfo(
-      selectedQuery,
-      databaseInfo,
-      quickEval,
-      range,
-    );
-    const item = new LocalQueryInfo(initialInfo, source);
-    qhm.addQuery(item);
     try {
-      const completedQueryInfo = await qs.compileAndRunQueryAgainstDatabase(
+      token.onCancellationRequested(() => source.cancel());
+
+      const localQueryRun = await this.createLocalQueryRun(
+        selectedQuery,
         databaseItem,
-        initialInfo,
-        queryStorageDir,
-        progress,
-        source.token,
-        undefined,
-        item,
-      );
-      qhm.completeQuery(item, completedQueryInfo);
-      await showResultsForCompletedQuery(
-        localQueryResultsView,
-        item as CompletedLocalQueryInfo,
-        WebviewReveal.Forced,
+        coreQueryRun.outputDir,
+        source,
       );
-      // Note we must update the query history view after showing results as the
-      // display and sorting might depend on the number of results
-    } catch (e) {
-      const err = asError(e);
-      err.message = `Error running query: ${err.message}`;
-      item.failureReason = err.message;
-      throw e;
+
+      try {
+        const results = await coreQueryRun.evaluate(
+          progress,
+          source.token,
+          localQueryRun.logger,
+        );
+
+        await localQueryRun.complete(results);
+
+        return results;
+      } catch (e) {
+        // It's odd that we have two different ways for a query evaluation to fail: by throwing an
+        // exception, and by returning a result with a failure code. This is how the code worked
+        // before the refactoring, so it's been preserved, but we should probably figure out how
+        // to unify both error handling paths.
```

Comment:
I think this is because it depends on where the error happens. If the error happens in the query server, then we get a result with a failure code. It can also happen that the error happens in the IDE, in which case, we need to handle the thrown exception.

It would be nice to figure out a way to unify.

Reply:
I think it's probably easy enough to unify, but I didn't want to attempt that in this PR.